libraries/SoaringController/SoaringController.cpp

```diff
@@ -0,0 +1,415 @@
+    // @Range: 0 0.5
+    // @User: Advanced 
+    AP_GROUPINFO("POLAR_B", 9, SoaringController, polar_B, 0.031),
+    
+    // @Param: POLAR_K
+    // @DisplayName: Cl factor
+    // @Description: Cl factor 2*m*g/(rho*S)
+    // @Units: m^2 s^-2
+    // @Range: 0 0.5
+    // @User: Advanced 
+    AP_GROUPINFO("POLAR_K", 10, SoaringController, polar_K, 25.6),
+    
+    // @Param: ALT_MAX
+    // @DisplayName: Maximum soaring altitude
+    // @Description: Don't thermal any higher than this.
+    // @Units: m
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_MAX", 11, SoaringController, alt_max, 350.0),
+    
+    // @Param: ALT_MIN
+    // @DisplayName: Minimum soaring altitude
+    // @Description: Don't get any lower than this.
+    // @Units: m
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_MIN", 12, SoaringController, alt_min, 50.0),
+    
+    // @Param: ALT_CUTOFF
+    // @DisplayName: Maximum power altitude
+    // @Description: Cut off throttle at this alt.
+    // @Units: m
+    // @Range: 0 1000.0
+    // @User: Advanced 
+    AP_GROUPINFO("ALT_CUTOFF", 13, SoaringController, alt_cutoff, 250.0),
+
+    AP_GROUPEND
+};
+    
+void SoaringController::get_target(Location &wp)
+{
+    wp=_prev_update_location;
+    location_offset(wp,ekf.X[2],ekf.X[3]);
+}
+bool SoaringController::suppress_throttle()
+{
+    if (_throttle_suppressed && (_alt<alt_min)) {
+        // Time to throttle up
+        _throttle_suppressed=false;
+    }
+    else if ((!_throttle_suppressed) && (_alt>alt_cutoff)) {
+        // Start glide
+        _throttle_suppressed=true;
+        // Zero the pitch integrator - the nose is currently raised to climb, we need to go back to glide.
+        _spdHgt->reset_pitch_I();
+        _cruise_start_time_us = AP_HAL::micros64();
+        // Reset the filtered vario rate - it is currently elevated due to the climb rate and would otherwise take a while to fall again,
+        // leading to false positives.
+        _filtered_vario_reading = 0;
+    }
+          
+    return _throttle_suppressed;
+}
+bool SoaringController::get_throttle_suppressed()
+{
+    return _throttle_suppressed;
+}
+
+void SoaringController::set_throttle_suppressed(bool suppressed)
+{
+    _throttle_suppressed = suppressed;
+}
+
+bool SoaringController::check_thermal_criteria()
+{
+    return(soar_active && ((AP_HAL::micros64() - _cruise_start_time_us ) > ((unsigned)min_cruise_s * 1e6)) && _filtered_vario_reading > thermal_vspeed && _alt < alt_max && _alt>alt_min);
+}
+bool SoaringController::check_cruise_criteria()
+{
+    float thermalability = (ekf.X[0]*exp(-pow(_loiter_rad/ekf.X[1],2)))-EXPECTED_THERMALLING_SINK; 
+    if (soar_active && (AP_HAL::micros64() -_thermal_start_time_us) > ((unsigned)min_thermal_s * 1e6) && thermalability < McCready(_alt)) {
+        hal.console->printf("Thermal weak, recommend quitting: W %f R %f th %f alt %f Mc %f\n",ekf.X[0],ekf.X[1],thermalability,_alt,McCready(_alt));
```

Comment:
I wasn't aware of the routines you mentioned. Does gcs_send_text do exactly the same thing as printf? For now, I've left this as is in PR #5737.